Add a check to PUT /user/:id to disallow name edits if an SSO user

adam-james-v opened the pull request:
Fixes: https://github.com/metabase/metabase/issues/16315
(since frontend changes to the UI already address the same concerns)

The ability for an SSO user to edit their names should be disallowed via API to mirror the same behaviour on the frontend.

adam-james-v committed:
Add a check to PUT /user/:id to disallow name edits if an SSO user

adam-james-v committed:
Clean up After SAML SSO tests

The `:sso_source` key is set for the Rasta user in some SAML tests, but is expeted to be nil in subsequent tests, so
we clean up in the SAML test ns.

adam-james-v committed:
Add a test to ensure SSO user names can't be changed via API

adam-james-v committed:
Missed a change I had made while adjusting tests

escherize commented on src/metabase/api/user.clj:
```diff
@@ -301,6 +315,11 @@
+    (when (or (contains? body :first_name)
+              (contains? body :last_name))
+      (api/checkp (valid-name-update? user-before-update first_name last_name)
+        "name" (tru "Editing names is not allowed for SSO users.")))
```
This call to `api/checkp` has `"name"` as its `field-name`. In this case, `field-name` can be used to highlight the form field that caused the an error in the FE. But since `"name"` isnt the name of one of the fields, that won't happen. 

Here it is with "name" as the `field-name`:
<img width="317" alt="Screen Shot 2022-07-07 at 3 36 27 PM" src="https://user-images.githubusercontent.com/343288/177876409-040c1abf-d642-4883-b311-0b18eba704ca.png">

Then again with "first name":
<img width="545" alt="Screen Shot 2022-07-07 at 3 36 50 PM" src="https://user-images.githubusercontent.com/343288/177876472-3d434bc7-b8c6-4994-bd9e-987beb5fcbf0.png">

I think highlighting the invalid field on FE in red only happens when `field-name` matches up with the specified field.

So you may want to split this up into 2 checks, one for `"first_name"` and one for `"last_name"`. 

Maybe something like:


```clojure
(when (contains? body :first_name)
  (api/checkp (valid-name-update? user-before-update first_name)
    "first_name" (tru "Editing first name is not allowed for SSO users.")))
(when (contains? body :last_name)
  (api/checkp (valid-name-update? user-before-update last_name)
    "last_name" (tru "Editing last name is not allowed for SSO users.")))
```

Of course you'll need to update `valid-name-update?` a bit too.

adam-james-v replied:
Hmm, this is a great point. I'll make the change. Though, It is also the case that this shouldn't actually occur. We have (or are supposed to have) made a change that prevents the UI from even showing those fields when a user is an SSO user. Still, a worthwhile improvement, thanks :)

escherize reviewed (commented):
I think its worth showing the user which field caused the error. So then the proposed change (from name to first_name and last_name) is worth doing.

adam-james-v committed:
valid-name-update? take 1 name, to allow better error messages

Let's the user know that first or last name is the cause of a problem, rather than just 'names'.

metamben commented on test/metabase/api/user_test.clj:
```diff
@@ -641,6 +642,29 @@
+                                         :sso_source   "jwt"
+                                         :is_superuser true}]
+        (letfn [(change-user-via-api! [m]
+                  (-> (mt/user-http-request :crowberto :put 400 (str "user/" user-id) m)))]
```
```suggestion
                  (mt/user-http-request :crowberto :put 400 (str "user/" user-id) m))]
```

metamben commented on test/metabase/api/user_test.clj:
```diff
@@ -641,6 +642,29 @@
+            (is (= {:errors {:last_name "Editing last name is not allowed for SSO users."}}
+                   (change-user-via-api! {:last_name "USER"}))))
+          (testing "New names that are the same as existing names succeed because there is no change."
+            (is (= {:first_name "SSO"
+                    :last_name  "User"}
+                   (-> (mt/user-http-request :crowberto :put 200 (str "user/" user-id) {:first_name "SSO"
+                                                                                        :last_name  "User"})
+                       (select-keys [:first_name :last_name]))))))))))
```
Might use `partial=` instead of `=` and `select-keys`, I think.

adam-james-v replied:
where is partial= defined? I can't get it to resolve in my REPL?

metamben replied:
Have you tried putting into the `is` check instead of `=`?

metamben reviewed (approved):
LGTM, but might want to wait for @escherize to bless it.

adam-james-v committed:
Remove unneeded thread macro

adam-james-v committed:
Use partial=

metamben commented on test/metabase/api/user_test.clj:
```diff
@@ -641,6 +642,28 @@
+          (testing "New names that are the same as existing names succeed because there is no change."
+            (is (partial= {:first_name "SSO"
+                           :last_name  "User"}
+                          (mt/user-http-request :crowberto :put 200 (str "user/" user-id) {:first_name "SSO"
+                                                                                           :last_name  "User"})))))))))
```
Adding the response code as parameter for `hange-user-via-api!` would let you reuse it here too, I think.

adam-james-v committed:
slight change to local fn for reusability
